models/repo/star.go

```
@@ -24,7 +25,12 @@ func init() {
 }
 
 // StarRepo or unstar repository.
+//
+// Will do nothing if stars are disabled.
 func StarRepo(ctx context.Context, doer *user_model.User, repo *Repository, star bool) error {
+	if setting.Repository.DisableStars {
```

It seems we just need to check that on router level.

---

So the reason why I added this here is because [some actions](https://github.com/go-gitea/gitea/blob/c49823c2b6dd4dca46f56e1c4b380d041c6edeb4/routers/web/repo/repo.go#L321) bypass the API and use these model functions directly, so instead of adding this check to every single one (and risk missing any, specially if added later) I simply changed the model functions themselves.

---

Perhaps we could return an err? As then the error will be floated up to the routes.

I agree that this should be five in the model, as then we could have the case of before where something might be disabled in the ui, but then still continues to work in the api.

---

> Perhaps we could return an err? As then the error will be floated up to the routes.

No, it will cause more problems if the err can't be properly handled. Golang's error system is not that modern, Gitea will report 500 errors for unhandled errors.

---

I returned empty values instead of errors because it was easier as I'm not quite sure what kind of error to return there, and because IsStaring only returns a bool, so it made sense to be consistent but I'll happily change the others to return errors instead if you guys think it's better.

---

> > Perhaps we could return an err? As then the error will be floated up to the routes.
> 
> 
> 
> No, it will cause more problems if the err can't be properly handled. Golang's error system is not that modern, Gitea will report 500 errors for unhandled errors.

Fair enough.

---

If it can not be checked on router level, it should be checked in the handler function, not in the model.

---

> not in the model, or there will be plenty of if conditions in the code, and most of them are duplicated.

but wouldn't moving the checks to the model functions which are used in multiple places _reduce_ the amount of checks in the rest of the code?

---

`model` is quite low level.
> cmd -> routers -> services -> models -> modules
> https://docs.gitea.com/contributing/guidelines-backend#package-dependencies

If these checks are only in models, you need to add error handling process in higher level which may introduce more checking codes. Why error handling process is neccesary? Actually, it should not return `nil` error here, as from the high level function, they will only get `nil` (error) returned. They don't know what happened, and generally will consider it as sucessed, but actually it is blocked. So for high level functions how to return the correct http status? 200 or 403/404?
That's the reason why I said

> If it can not be checked on router level, it should be checked in the route handle function or services level

which means these check should be in high level (routers/services) not in low level.
This is what I’ve considered the design of this project, it might not be correct, if it is, please tell me.